File: shared/pnl.py
from __future__ import annotations


def compute_pnl_usdt(
    *,
    entry: float,
    quantity: float,
    exit_price: float | None,
    side: str = "BUY",
) -> float | None:
    if exit_price is None or entry <= 0 or quantity <= 0:
        return None
    if side.upper() == "BUY":
        return round((exit_price - entry) * quantity, 4)
    return round((entry - exit_price) * quantity, 4)


def compute_pnl_pct(
    *,
    entry: float,
    exit_price: float | None,
    side: str = "BUY",
) -> float | None:
    if exit_price is None or entry <= 0:
        return None
    if side.upper() == "BUY":
        return round((exit_price - entry) / entry * 100, 4)
    return round((entry - exit_price) / entry * 100, 4)


def compute_pnl_usdt_directional(
    *,
    entry: float,
    quantity: float,
    exit_price: float | None,
    direction: str,
) -> float | None:
    side = "BUY" if direction.upper() == "LONG" else "SELL"
    return compute_pnl_usdt(
        entry=entry, quantity=quantity, exit_price=exit_price, side=side,
    )


def compute_pnl_pct_directional(
    *,
    entry: float,
    exit_price: float | None,
    direction: str,
) -> float | None:
    side = "BUY" if direction.upper() == "LONG" else "SELL"
    return compute_pnl_pct(entry=entry, exit_price=exit_price, side=side)

File: frontend/src/lib/pnl.ts
export function computePnlUsdt(
  entry: number,
  quantity: number,
  exitPrice: number | null | undefined,
  side = "BUY",
): number | null {
  if (exitPrice == null || entry <= 0 || quantity <= 0) return null;
  if (side.toUpperCase() === "BUY") return Math.round((exitPrice - entry) * quantity * 10000) / 10000;
  return Math.round((entry - exitPrice) * quantity * 10000) / 10000;
}

export function computePnlPct(
  entry: number,
  exitPrice: number | null | undefined,
  side = "BUY",
): number | null {
  if (exitPrice == null || entry <= 0) return null;
  if (side.toUpperCase() === "BUY") return Math.round(((exitPrice - entry) / entry) * 100 * 10000) / 10000;
  return Math.round(((entry - exitPrice) / entry) * 100 * 10000) / 10000;
}

export function fmtPnlValue(
  pnlUsdt: number | null | undefined,
  pnlPct: number | null | undefined,
  decimals = 2,
): string {
  if (pnlUsdt == null) return "—";
  const sign = pnlUsdt >= 0 ? "+" : "";
  const amount = `${sign}$${pnlUsdt.toLocaleString("es-AR", {
    minimumFractionDigits: decimals,
    maximumFractionDigits: decimals,
  })}`;
  if (pnlPct == null) return amount;
  const pctSign = pnlPct >= 0 ? "+" : "";
  return `${amount} (${pctSign}${pnlPct.toFixed(2)}%)`;
}

export function pnlColorClass(pnl: number | null | undefined): string {
  if (pnl == null) return "text-zinc-400";
  if (pnl > 0) return "text-emerald-400";
  if (pnl < 0) return "text-red-400";
  return "text-zinc-400";
}

export type PositionSide = "LONG" | "SHORT";

export function tradeDirection(trade: { position_side?: string; side?: string }): PositionSide {
  if (trade.position_side === "SHORT") return "SHORT";
  if (trade.side === "SELL" && trade.position_side !== "LONG") return "SHORT";
  return "LONG";
}

export function computePnlUsdtDirectional(
  entry: number,
  quantity: number,
  exitPrice: number | null | undefined,
  direction: PositionSide,
): number | null {
  return computePnlUsdt(entry, quantity, exitPrice, direction === "SHORT" ? "SELL" : "BUY");
}

export function computePnlPctDirectional(
  entry: number,
  exitPrice: number | null | undefined,
  direction: PositionSide,
): number | null {
  return computePnlPct(entry, exitPrice, direction === "SHORT" ? "SELL" : "BUY");
}

export function slDistancePct(
  entry: number,
  stopLoss: number,
  direction: PositionSide,
): number | null {
  if (entry <= 0) return null;
  if (direction === "SHORT") return ((stopLoss - entry) / entry) * 100;
  return ((entry - stopLoss) / entry) * 100;
}

export function tpDistancePct(
  entry: number,
  takeProfit: number,
  direction: PositionSide,
): number | null {
  if (entry <= 0) return null;
  if (direction === "SHORT") return ((entry - takeProfit) / entry) * 100;
  return ((takeProfit - entry) / entry) * 100;
}
